app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -7,7 +9,29 @@ def google_oauth2
     email = access_token.info['email']
 
     user = User.find_by(email:)
-    if !user
+
+    if user
+      # https://cyberinsider.com/unfixed-google-oauth-flaw-exposes-millions-to-account-takeovers/
+      sub = access_token.dig('extra', 'id_info', 'sub').presence!
+
+      if user.google_sub
+        if sub != user.google_sub
+          Rails.error.report(
+            Error.new(SubMismatch),
+            context: {
+              user_sub_in_db: user.google_sub,
+              sub_in_google_response: sub,
+            },
+          )
+
+          flash[:alert] = 'You are attempting a domain identity takeover attack. Blocked!'
+          redirect_to(new_user_session_path)
+          return
+        end
+      else
+        user.update!(google_sub: sub)
```

Comment:
We will gradually "backfill" users' Google `sub` values as they log in. Once a user logs in after this change has been deployed, then they should henceforth be protected from this identity takeover attack. Unfortunately, I don't think that there is a way to protect against this attack for users who don't log in after this change has been deployed.